### PR TITLE
Add OP_MULPOW2.

### PR DESCRIPTION
**Pull request by EyeOfPython**
This computes x n -> x * 2^n, which is useful for many applications.

### CODE REVIEW

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -93,7 +93,6 @@ static bool IsOpcodeDisabled(opcodetype opcode, uint32_t flags) {
         case OP_2MUL:
         case OP_2DIV:
         case OP_MUL:
-        case OP_RSHIFT:
```
Comment (schancel) [inline src/script/interpreter.cpp:96]: Why's this going away?

Comment (schancel) [reply src/script/interpreter.cpp:96]: Nevermind

Comment (EyeOfPython) [reply src/script/interpreter.cpp:96]: 😉

File: src/script/script.h [hunk 2]
```diff
@@ -362,6 +362,34 @@ class CScriptNum {
         return *this;
     }
 
+    inline CScriptNum mulpow2(const CScriptNum rhs) {
+        const int64_t sign = m_value > 0 ? 1 : -1;
+        // We use unsigned integers for shifts, they have well-defined behavior
+        const uint64_t absval = m_value > 0 ? m_value : -m_value;
+        if (absval == 0) {
+            return CScriptNum(0);
```
Comment (schancel) [inline src/script/script.h:370]: You can move this up first. Saves two branches in this case.

File: src/script/script.h [hunk 3]
```diff
@@ -362,6 +362,34 @@ class CScriptNum {
         return *this;
     }
 
+    inline CScriptNum mulpow2(const CScriptNum rhs) {
+        const int64_t sign = m_value > 0 ? 1 : -1;
+        // We use unsigned integers for shifts, they have well-defined behavior
+        const uint64_t absval = m_value > 0 ? m_value : -m_value;
```
Comment (schancel) [inline src/script/script.h:368]: Should note that this is defined as `x` for reference below.

**Review by schancel — CHANGES_REQUESTED** [review]

File: src/test/op_mulpow2_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        bool r = EvalScript(stack, script, flags, sigchecker, &err);
+        BOOST_CHECK_MESSAGE(
+            !r, "Expected " << ScriptErrorString(expected_error) << " but got "
+                            << ScriptErrorString(err) << " for '"
+                            << ScriptToAsmStr(script) << "'");
+        BOOST_CHECK_EQUAL(err, expected_error);
+    }
+}
+
+static bool Pow2(int64_t base, int64_t exponent, int64_t &result) {
```
Comment (schancel) [inline src/test/op_mulpow2_tests.cpp:78]: I would consider using the existing pow math functions. But this also works.

**Review by schancel — CHANGES_REQUESTED** [review]

**Review by schancel — APPROVED** [review]

**Pull request merged**